### PR TITLE
Split gas tests

### LINKED ISSUE #10: Decouple HexSumTree gas analysis from tests
Currently running `test/hex-tree.js` takes a lot of time given that we are adding a lot of entries to the tree to test the gas usage.

We should be doing this in a separate script and not in the unit tests.

### PR DESCRIPTION
**Pull request by bingen**
And add more gas tests for SumTree Sortition lib.

Fixes #10.

### CODE REVIEW

File: contracts/test/HexSumTreePublic.sol [hunk 1]
```diff
@@ -19,6 +19,10 @@ contract HexSumTreePublic {
         emit LogKey(bytes32(tree.insert(v)));
     }
 
+    function insertNoLog(uint256 v) public {
```
Comment (izqui) [inline contracts/test/HexSumTreePublic.sol]: Making these function external would be a bit more efficient which is better for more accurate gas benchmarks

File: test/hex-tree-gas.js [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    await tree.init()
+  })
+
+  const assertBN = (bn, n, m) => {
+    assert.equal(bn.toNumber(), n, m)
+  }
+
+  const logTreeState = async () => {
+    //console.log((await tree.getState()).map(x => x.toNumber()))
+    const state = await tree.getState()
```
Comment (izqui) [inline test/hex-tree-gas.js]:
```suggestion
    const [ depth, nextKey ] = await tree.getState()
```

File: test/hex-tree-gas.js [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    console.log(`Tree depth:    ${state[0]}`);
+    console.log(`Tree next key: ${state[1].toNumber().toLocaleString()}`);
```
Comment (izqui) [inline test/hex-tree-gas.js]: What's the reason for using `toLocaleString()`?

File: test/hex-tree-gas.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
+const assertRevert = require('./helpers/assert-revert')
+
+const HexSumTreePublic = artifacts.require('HexSumTreePublic')
+
+const CHILDREN = 16
+
+contract('Hex Sum Tree', (accounts) => {
```
Comment (izqui) [inline test/hex-tree-gas.js]:
```suggestion
contract('Hex Sum Tree (Gas analysis)', (accounts) => {
```

File: test/hex-tree-gas.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    console.log(`Tree depth:    ${state[0]}`);
+    console.log(`Tree next key: ${state[1].toNumber().toLocaleString()}`);
+    console.log(`Tree total sum: `, (await tree.totalSum()).toNumber().toLocaleString())
+  }
+
+  const logSortitionGas = async (value) => {
+    console.log(`Sortition ${value} gas:`, (await tree.sortition.estimateGas(value)).toLocaleString())
+  }
+
+  const logSortitionHex = async (value) => {
```
Comment (izqui) [inline test/hex-tree-gas.js]: Function is unused

File: contracts/lib/HexSumTree.sol [hunk 6]
```diff
@@ -86,7 +86,7 @@ library HexSumTree {
         uint256 checkingNode;
         uint256 nodeSum;
         for (; checkingLevel > INSERTION_DEPTH; checkingLevel--) {
-            for (; child < CHILDREN; child++) {
+            for (child = 0; child < CHILDREN; child++) {
```
Comment (izqui) [inline contracts/lib/HexSumTree.sol]: How weren't we having issues with this before?

File: test/hex-tree-gas.js [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    console.log('|---------|' + '-'.padStart(COL_SIZE + 2, '-') + '|' + '-'.padStart(COL_SIZE + 2, '-') + '|')
+    console.log('| Min     |', formatDivision(min / batchSize, COL_SIZE), '|', formatDivision((min - TX_GAS) / batchSize, COL_SIZE), '|')
+    console.log('| Max     |', formatDivision(max / batchSize, COL_SIZE), '|', formatDivision((max - TX_GAS) / batchSize, COL_SIZE), '|')
+    console.log('| Average |', formatDivision(avg / batchSize, COL_SIZE), '|', formatDivision((avg - TX_GAS) / batchSize, COL_SIZE), '|')
+    console.log()
+  }
+
+  it('inserts one node', async () => {
+    const r = await tree.insertNoLog(10)
+    const insertGas = r.receipt.cumulativeGasUsed
```
Comment (izqui) [inline test/hex-tree-gas.js]: We should use `gasUsed`, `cumulativeGasUsed` includes the gas used by other transactions in the block that came before this one. This is not an issue if there is always just one transaction per block, but we should fix it.

File: test/hex-tree-gas.js [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+  const logTreeState = async () => {
+    //console.log((await tree.getState()).map(x => x.toNumber()))
+    const state = await tree.getState()
+    console.log(`Tree depth:    ${state[0]}`);
+    console.log(`Tree next key: ${state[1].toNumber().toLocaleString()}`);
+    console.log(`Tree total sum: `, (await tree.totalSum()).toNumber().toLocaleString())
+  }
+
+  const logSortitionGas = async (value) => {
+    console.log(`Sortition ${value} gas:`, (await tree.sortition.estimateGas(value)).toLocaleString())
```
Comment (izqui) [inline test/hex-tree-gas.js]: I have done a few tests and using `estimateGas` seems to return the exact same value as the gas used by performing the call as a transaction.

File: test/hex-tree-gas.js (hunk 4 above)
Comment (izqui) [inline test/hex-tree-gas.js]: This test file is quite slow and I don't see any assertions that aren't done somewhere else. We should consider skipping these tests in most cases, unless you are explicitly profiling gas.

File: test/hex-tree-gas.js [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+  }
+
+  const formatDivision = (result, colSize) => {
+    return Math.round(result).toLocaleString().padStart(colSize, ' ')
+  }
+  const logGasStats = (title, gasArray, batchSize = 1) => {
+    const COL_SIZE = 7
+    console.log(title)
+    console.log('Size:   ', gasArray.length)
+    const TX_GAS = 21000
```
Comment (izqui) [inline test/hex-tree-gas.js]: Not a really big difference, but if we wanted to truly know how much gas a function uses we could use something like this:

```sol
modifier profileGas {
    uint256 initialGas = gasleft();
    _;
    emit GasConsumed(initialGas - gasleft());
  }
```

This is the most accurate way I can think of to properly profile how much gas a function really uses.

File: test/hex-tree-gas.js (hunk 8 above)
Comment (bingen) [reply test/hex-tree-gas.js]: Yes, I did some tests too and had the same result. As rule of thumb, I'm using `estimateGas` when I don't need the transaction to be executed (e.g., when it's a read call), as the code is shorter, and the real call and the receipt when I need the state change.

File: test/hex-tree-gas.js (hunk 5 above)
Comment (bingen) [reply test/hex-tree-gas.js]: Yes, but it's so useful to have it handy when things go wrong! ;-) Besides, as we talked offline, these are not going to be the final tests, so meanwhile I would like to keep it.

File: test/hex-tree-gas.js (hunk 3 above)
Comment (bingen) [reply test/hex-tree-gas.js]: To have thousands separators. Is there a better way of doing it?

File: test/hex-tree-gas.js (hunk 8 above)
Comment (izqui) [reply test/hex-tree-gas.js]: That makes sense. Gas estimations tend to break when there are refunds

File: contracts/lib/HexSumTree.sol (hunk 6 above)
Comment (bingen) [reply contracts/lib/HexSumTree.sol]: Added test for it: a6559ab

**Pull request merged**